### ISSUE TITLE
Type inference sometimes fails inside method for cast class attribute

### ISSUE THREAD

**Issue by pydsigner** [microsoft/pyright#10847]
Type: <b>Bug</b>

The type inference behavior of a cast class attribute appears to be irregular. If we take these two classes in this minimal reproducing example:

```py
import typing
from collections import Counter
from kivy.app import App


class Test:
    test_attr = typing.cast(App, object())
    def func(self):
        self.test_attr.build()

class Test2:
    test_attr = typing.cast(Counter, object())
    def func(self):
        self.test_attr.get(1)
```
The expected behavior is for `Test.test_attr` to be inferred as kivy.App and `Test2.test_attr` to be inferred as `Counter` inside `Test.func` and `Test2.func`, respectively. However, `Test.test_attr` is being inferred as `Unknown`, while `Test2.test_attr` infers correctly.

Python version: 3.13.2
Extension version: 2025.7.1
VS Code version: Code 1.103.0 (e3550cfac4b63ca4eafca7b601f0d2885817fd1f, 2025-08-06T21:40:10.271Z)
OS version: Windows_NT x64 10.0.26100
Modes:

**Comment by rchiodo** [comment]
`kivy` is on Pypi. I can reproduce this as well.

**Comment by erictraut** [comment]
Pyright is working as intended here. The problem is that the `App` symbol imported from `kivy` is implemented in an extension module (i.e. a module written in a language other than Python), and there is no stub file that defines its type. Pyright therefore infers its type as `Unknown`. To fix this, you will need to either provide your own stub file or work with the maintainer of the `kivy` library to include a stub file with the library. For more information about stub files, refer to [this documentation](https://microsoft.github.io/pyright/#/type-stubs).

**Comment by pydsigner** [comment]
@erictraut I see a potentially related issue in #6096. However, I think it ought to be at least documented that `typing.cast` is a no-op for `Unknown` types, if it's not going to do the expected thing and actually convert the type. Further, if the checker essentially suppresses the cast, actually emitting messaging along those lines instead of producing head-scratching errors downstream in the process would be very useful.
